src/main/scala/mesosphere/marathon/state/Container.scala

```diff
@@ -3,20 +3,31 @@ package mesosphere.marathon.state
 import com.wix.accord.dsl._
 import com.wix.accord._
 import mesosphere.marathon.api.v2.Validation._
-import org.apache.mesos.{ Protos => Mesos }
+
+import org.apache.mesos.{ Protos => mesos }
+
 import scala.collection.immutable.Seq
 
-// TODO: trait Container and specializations?
-// Current implementation with type defaulting to DOCKER and docker to NONE makes no sense
-case class Container(
-    `type`: Mesos.ContainerInfo.Type = Mesos.ContainerInfo.Type.DOCKER,
-    volumes: Seq[Volume] = Nil,
-    docker: Option[Container.Docker] = None) {
+case class Container(`type`: mesos.ContainerInfo.Type,
```

Comment:
The original idea to refactor the Container model was more like this (simplified):
```
sealed trait Container {
  def volumes: Seq[Volume]
}

case class Docker(volumes: Seq[Volume]) extends Container
case class Mesos(volumes: Seq[Volume]) extends Container
case class Universal(volumes: Seq[Volume]) extends Container
```

This would let us get rid of things like _hasConsistentType_, lots of overloaded constructors, and having a mesos.Protos.ContainerInfo.Type as a member of the Container. Plus, it would make constructing a container based on a proto much easier, and type safe, as each specific case class implementing the Container trait would require a specific set of properties.

Did you consider this approach?